### PR TITLE
Multi-Dataset Validation (LM-Loss/Perplexity)

### PR DESCRIPTION
**Pull request by bigximik**
# ✨ Description

Phase 1 of #65

Closes # 

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [ ] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. Change A
2. Change B

## ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General

- [ ] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/developers/contributing).
- [ ] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [ ] 🎉 The functionality is complete, and I have tested the changes.
- [ ] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration

- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing

- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

**Comment by bigximik** [comment]
@jlamypoirier What do you think of such approach?

File: fast_llm/data/data/gpt/data.py [hunk 1]
```diff
@@ -91,18 +91,39 @@ def setup(
...
-                    cache_directory=self._cache_directory,
-                    distributed=distributed,
-                    phase=phase,
-                    sequence_length=self._max_sequence_length,
-                    vocab_size=self._vocab_size,
-                    tokenizer=self._tokenizer,
-                )
-                dataset = self._config.datasets[phase].build_and_sample(sampling)
-                self._datasets[phase] = DatasetMonitor(dataset, self._config.data_sample_warn_time_ms)
+                if phase == PhaseType.validation:
```
Comment (jlamypoirier) [inline fast_llm/data/data/gpt/data.py:94]: Doesn't seem like all that extra handling for validation is necessary. The `Data` doesn't need to know about phases anyway, how about we just use `datasets: dict[str, GPTSampledDatasetConfig]` and let the trainer worry about phases?

File: fast_llm/engine/training/trainer.py [hunk 2]
```diff
@@ -184,7 +184,7 @@ def _train(self) -> tuple[bool, dict[PhaseType, dict[str, typing.Any]]]:
             self._completed_steps,
             self._config.training.prefetch_factor,
         )
-        valid_iterator = None
+        valid_iterators = {name: None for name in self._data.validation_dataset_names}
```
Comment (jlamypoirier) [inline fast_llm/engine/training/trainer.py:187]: I think we need to have a separate validation config for each validation dataset. That would take care of defining which datasets are for validation, allow setting a different number of iteration depending on the dataset (and a different interval if we really want to), and we'll need if we plan to go ahead with custom evaluators.

(I'm also thinking the dataset config should go in there, but it can wait.)

Comment (bigximik) [reply fast_llm/engine/training/trainer.py:187]: do you mean something like this:
```python
TrainerConfig:
  validation: <ValidationsConfig>
    common_config: ValidationConfig
    datasets: <dicts of name: ValidationInfo>
        name:
            dataset: GPTSampledDatasetConfig
            config: ValidationConfig | None
```

File: fast_llm/data/data/gpt/data.py (hunk 1 above)
Comment (bigximik) [reply fast_llm/data/data/gpt/data.py:94]: Would we then need to move the `torch.DataLoader` encapsulation from `GptData` to `Trainer`, given that the format of training and validation dataset configurations will be different? This is especially relevant if we introduce a per-validation dataset `ValidationConfig`.

Comment (jlamypoirier) [reply fast_llm/data/data/gpt/data.py:94]: Why would it be different? The `DataLoader` part would be unchanged, the only difference on the data side would be to swap `PhaseType` for `str`

File: fast_llm/engine/training/trainer.py (hunk 2 above)
Comment (jlamypoirier) [reply fast_llm/engine/training/trainer.py:187]: Not really, I'm suggesting something like:
```
training: <TrainingConfig>
  validation: <dict[str, ValidationConfig]>
    dataset_1: <ValidationConfig>
      interval: 500
      iterations: 10      
    dataset_2: <ValidationConfig>
      interval: 500
      iterations: 20      
    dataset_3: <ValidationConfig>
      interval: 1000
      iterations: 5    
```
That would be enough for now.

In the future we'll probably want to move dataset configs here and have some kind of default to reduce verbosity for common parameters, ex. use the same interval for all (something like #168, #171), but let's leave that out of the present PR.

File: fast_llm/data/data/gpt/data.py (hunk 1 above)
Comment (bigximik) [reply fast_llm/data/data/gpt/data.py:94]: how to distinguish  phases then?  to have a composite key, something like `training:dataset_name1`, `validation:dataset_name1`, `validation:dataset_name2`?

Comment (jlamypoirier) [reply fast_llm/data/data/gpt/data.py:94]: We don't. The `Data` doesn't need to know about phase, they are handled in the trainer.

Comment (jlamypoirier) [reply fast_llm/data/data/gpt/data.py:94]: (Only exception is the legacy dataset format which is the reason we have `GPTSamplingData.phase`, but it's deprecated so a simple hack will do.)

**Comment by bigximik** [comment]
I have implemented it so that the config will look something like this:
```yaml
data:
  datasets:
    Training: # training dataset, hardcoded name
      type: memmap
      path: some_path1
    Test:  # test dataset, hardcoded name
      type: memmap2
      path: some_path2
    validation_dataset_name1:  # validation dataset, any name
      type: memmap
      path: some_path3
    validation_dataset_name2:  # validation dataset, any name
      type: memmap
      path: some_path4

training:
  training_iters: 2
  test_iters: 2
  validation:
    validation_dataset_name1:
      interval: 2
      iterations: 1
    validation_dataset_name2:
      interval: 2
      iterations: 1
```
@jlamypoirier, have I got it right?

**Comment by bigximik** [comment]
Maybe to add `training_dataset_name` and `test_dataset_name` into `TrainingConfig` with default values `Training` and `Test` so people can override it if they like?

**Comment by jlamypoirier** [comment]
> I have implemented it so that the config will look something like this:
> @jlamypoirier, have I got it right?

Yes

> Maybe to add `training_dataset_name` and `test_dataset_name` into `TrainingConfig` with default values `Training` and `Test` so people can override it if they like?

I don't think it's worth it, but feel free to fix the capitalization if you find a way to do it without breaking backward comparibility

File: fast_llm/engine/training/config.py [hunk 3]
```diff
@@ -267,9 +267,9 @@ class ShutdownConfig(IntervalConfig):
 
 @config_class()
 class TrainingConfig(Config):
-    validation: ValidationConfig = Field(
-        default_factory=ValidationConfig,
-        desc="Configuration for the validation phase",
+    validation: dict[str, ValidationConfig] = Field(
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/training/config.py:270]: I decided not to rename field at this time as it brings other questions like do we rename `ValidationConfig` to `EvaluationConfig` as well, do we name the phase `Validation` at all if all the rest be named `Evaluation`. Also, do we name fields in metrics `Evaluation.dataset_name.metric` instead of `Validation.dataset_name.metric`. 

However i have added backward compatibility. But propose to rename `validation` to `evaluations` during the next phase, then we will decide on all the chain of renamings.

What do you think?

File: docs/quick-start.md [hunk 4]
```diff
@@ -493,8 +493,9 @@ Save the following as `fast-llm-tutorial/train-config.yaml`:
       logs:
         interval: 10
       validation:
-        iterations: 25
-        interval: 100
+        Validation:
```
[the remark this answers is not in the document]
Comment (bigximik) [reply docs/quick-start.md:496]: I have implemented this and will update the tests and documentation along with the fixes for the above issues.

File: fast_llm/data/data/gpt/data.py [hunk 5]
```diff
@@ -97,23 +97,23 @@ def setup(
             warnings.warn(f"Using the dataset directory for the index cache.")
 
         self._datasets = {}
-        for phase, num_samples in samples_per_phase.items():
+        for dataset_name, num_samples in samples_per_dataset.items():
             if num_samples > 0:
                 # TODO: Do the check earlier.
-                assert phase in self._config.datasets
+                assert dataset_name in self._config.datasets
```
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/data/data/gpt/data.py]: added warning

File: fast_llm/engine/training/config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bigximik) [reply fast_llm/engine/training/config.py]: renamed to evaluations and renamed `ValidationConfig` to `EvaluationConfig` left phase as Validation

File: docs/quick-start.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bigximik) [reply docs/quick-start.md]: updated test cases and docks

**Comment by bigximik** [comment]
I have changed `validation` to `evaluations` and `ValidationConfig` to `EvaluationConfig`, but kept the phase as `Validation`.  
So, in the documentation, it is still referred to as validation datasets, which need to be defined in `data.datasets`, while their usage should be specified in `training.evaluations`.

File: fast_llm/data/data/gpt/data.py [hunk 6]
```diff
@@ -77,18 +79,43 @@ def __init__(
...
         timeout: float | None = None,
     ) -> None:
         """
         Load the datasets, and prepare or load the samplings.
         This may take a while and a significant amount of cpu memory.
         """
-        super().setup(distributed, samples_per_phase, cache_directory)
+        # Some dataset names may come from phases and are capitalized,
+        # so we need to normalize them before use.
+        samples_per_dataset = {key.lower(): value for key, value in samples_per_dataset.items()}
```
Comment (jlamypoirier) [inline fast_llm/data/data/gpt/data.py:100]: Making things case-insensitive does a lot more than backward compatibility, how about we just do `PhaseType.[x].lower()` in training instead?
Either way, need to ensure that `lower()` doesn't create duplicated (ex. if `Validation` and `validation` are both defined)

File: fast_llm/data/data/gpt/data.py [hunk 7]
```diff
@@ -77,18 +79,43 @@ def __init__(
...
+
+        # Check and raise an error if a used dataset is not defined.
+        for dataset_name in samples_per_dataset.keys():
```
Comment (jlamypoirier) [inline fast_llm/data/data/gpt/data.py:103]: If we want to do these check during validation, we could also do the check in the training config, which has access to both the data and evaluation configs.

File: fast_llm/data/data/gpt/data.py [hunk 8]
```diff
@@ -77,18 +79,43 @@ def __init__(
         self._max_sequence_length = max_sequence_length
         self._cross_document_attention = cross_document_attention
 
+        # This is a mapping between the dataset name in the config and its normalized variant for later reference.
+        # This is needed because previously, dataset names were based on phases and were capitalized,
+        # but this is no longer a requirement.
+        self._dataset_name_map = {key.lower(): key for key in self._config.datasets.keys()}
```
Comment (jlamypoirier) [inline fast_llm/data/data/gpt/data.py:85]: How about we only rename known values (`Training`, `Validation`, `Test`) in `_from_dict` instead? 
(Also redundant with the `__post_init__` thing)

File: fast_llm/data/dataset/config.py [hunk 9]
```diff
@@ -40,10 +40,14 @@ class SamplingData:
     cache_directory: pathlib.Path | None
     # TODO: This prevents the sampling config from being pickled in multiprocessing.
     distributed: "Distributed"
-    phase: PhaseType
+    dataset_name: str
     # Using a mutable rather than an int so it's shared with all copies made with `update`.
     _rank_counter: typing.Iterator[int] = itertools.count
 
+    def __post_init__(self):
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/config.py:47]: `__post_init__` shouldn't be implemented. You need to use `_validate` instead, though for this kind of backward compatibility fix `_from_dict` is more appropriate.

**Review by jlamypoirier — COMMENTED** [review]
Almost ready to merge, some minor issues and suggestions

**Comment by tscholak** [comment]
Guys can we get to a conclusion here please? I’d like us to merge this by tomorrow end of day, at the latest. Thanks

**Review by jlamypoirier — APPROVED** [review]
@bigximik Adjusted the code, ready to merge if it works for you.

**Review by tscholak — APPROVED** [review]
thank you both, looks good to me!

File: fast_llm/data/data/gpt/data.py (hunk 6 above)
Comment (bigximik) [reply fast_llm/data/data/gpt/data.py]: Thanks, missed that one about duplicates like `Validation` vs `validation`

**Comment by bigximik** [comment]
I have updated the instruction fine-tuning documentation that came from the merge but created a separate issue for moving the check of used dataset definitions to `_validate` in `Trainer` (#213).
